File: src/Organon/Syl/Document.hs
module Organon.Syl.Document
  ( -- * Types
    Document (..),
    ProofBlock (..),
    Premise (..),
    RefModifier (..),
    SrcPos (..),
    Located (..),

    -- * Parsing
    parseDocument,
    fromConcreteH,
  )
where

import Data.Text (Text)
import qualified Data.Text as T
import Organon.Common.Types (Located (..), SrcPos (..))
import Organon.Syl.Hole (PropTypeH (..), PropositionH (..), TermH (..))
import Organon.Syl.Parser (Parser, isNameChar, lexeme, lineComment, propositionHP, symbol)
import Organon.Syl.Types
import Text.Megaparsec
import Text.Megaparsec.Char
import qualified Text.Megaparsec.Char.Lexer as L

-- | A modifier applied to a reference premise.
data RefModifier
  = -- | Simple conversion: swap subject and predicate (E↔E, I↔I).
    RefConv
  | -- | Conversion per accidens: swap terms and weaken (A→I, E→O).
    RefPerAccidens
  deriving stock (Eq, Ord, Show)

-- | A premise in a proof block: either a proposition, a reference, or
-- a proposition with holes.
data Premise
  = -- | A concrete proposition.
    PremiseProp Proposition
  | -- | A reference to a previously proved conclusion.
    --   Fields: optional namespace, name, optional conversion modifier.
    PremiseRef (Maybe Text) Text (Maybe RefModifier)
  | -- | A proposition containing holes (term, quantifier, or whole).
    PremiseHole PropositionH
  deriving stock (Eq, Show)

-- | A proof block.
data ProofBlock = ProofBlock
  { proofName :: Located Text,
    proofPremises :: [Located Premise],
    proofConclusion :: Located PropositionH
  }
  deriving stock (Eq, Show)

-- | A document: an optional tradition directive followed by proof blocks.
data Document = Document
  { docTradition :: Maybe (Located Tradition),
    docOpens :: [Located Text],
    docProofs :: [Located ProofBlock]
  }
  deriving stock (Eq, Show)

-- | Parse a .syl document.
parseDocument :: Text -> Either Text Document
parseDocument input =
  case parse (scn *> documentP <* eof) "" input of
    Left err -> Left (T.pack (errorBundlePretty err))
    Right d -> Right d

-- Whitespace including newlines (for between top-level blocks).
scn :: Parser ()
scn = L.space space1 lineComment empty

-- Position helpers

getPos :: Parser SrcPos
getPos = do
  sp <- getSourcePos
  pure (SrcPos (unPos (sourceLine sp)) (unPos (sourceColumn sp)))

located :: Parser a -> Parser (Located a)
located p = do
  s <- getPos
  v <- p
  e <- getPos
  pure (Located s e v)

-- Premise: either @Reference, a proposition with holes, or a concrete proposition.

premiseP :: Parser Premise
premiseP = refP <|> propPremiseP

propPremiseP :: Parser Premise
propPremiseP = do
  ph <- propositionHP
  case fromConcreteH ph of
    Just prop -> pure (PremiseProp prop)
    Nothing -> pure (PremiseHole ph)

refModifierP :: Parser (Maybe RefModifier)
refModifierP =
  optional (RefConv <$ symbol "conv" <|> RefPerAccidens <$ symbol "per-accidens")

refP :: Parser Premise
refP = do
  _ <- char '@'
  first <- lexeme (takeWhile1P (Just "proof name") isNameChar)
  mDot <- optional (char '.')
  case mDot of
    Just _ -> do
      name <- lexeme (takeWhile1P (Just "proof name") isNameChar)
      modifier <- refModifierP
      pure (PremiseRef (Just first) name modifier)
    Nothing -> do
      modifier <- refModifierP
      pure (PremiseRef Nothing first modifier)



-- Proof block

proofBlockP :: Parser ProofBlock
proofBlockP = do
  _ <- symbol "proof"
  name <- located (lexeme (takeWhile1P (Just "proof name") isNameChar))
  _ <- eol *> scn
  premises <- many (try (located premiseP <* eol <* scn))
  concl <- conclusionP
  pure (ProofBlock name premises concl)

conclusionP :: Parser (Located PropositionH)
conclusionP = do
  _ <- symbol "∴" <|> symbol "therefore"
  loc <- located propositionHP
  pure loc

-- | Extract a concrete Proposition from a PropositionH, if it has no holes.
fromConcreteH :: PropositionH -> Maybe Proposition
fromConcreteH (PropH (ConcretePT pt) (ConcreteT s) (ConcreteT p)) = Just (Proposition pt s p)
fromConcreteH _ = Nothing

-- Tradition directive

traditionP :: Parser (Located Tradition)
traditionP = located $ do
  _ <- symbol "tradition"
  choice
    [ Strict <$ symbol "Strict",
      Traditional <$ symbol "Traditional",
      Full <$ symbol "Full"
    ]

-- Open directives

openP :: Parser (Located Text)
openP = located $ do
  _ <- symbol "open"
  lexeme (takeWhile1P (Just "namespace name") isNameChar)

-- Document

documentP :: Parser Document
documentP = do
  trad <- optional (try (traditionP <* eol <* scn))
  opens <- many (try (openP <* eol <* scn))
  proofs <- many (located proofBlockP <* scn)
  pure (Document trad opens proofs)

File: web/src/core/document.ts
// Document parser — ported from Organon.Syl.Document

import type { PropositionH } from "./hole";
import { ParserState, isNameChar, propositionHP } from "./parser";
import { Proposition, PropType, Tradition } from "./types";
import { ParseError } from "./parser";

// ---------------------------------------------------------------------------
// Source positions & Located
// ---------------------------------------------------------------------------

export type SrcPos = {
  readonly posLine: number;
  readonly posCol: number;
};

export type Located<T> = {
  readonly locStart: SrcPos;
  readonly locEnd: SrcPos;
  readonly locValue: T;
};

// ---------------------------------------------------------------------------
// Premise types
// ---------------------------------------------------------------------------

export type RefModifier = "conv" | "per_accidens";

export type Premise =
  | { readonly tag: "PremiseProp"; readonly prop: Proposition }
  | {
      readonly tag: "PremiseRef";
      readonly namespace: string | null;
      readonly name: string;
      readonly modifier: RefModifier | null;
    }
  | { readonly tag: "PremiseHole"; readonly propH: PropositionH };

// ---------------------------------------------------------------------------
// Proof block & Document
// ---------------------------------------------------------------------------

export type ProofBlock = {
  readonly proofName: Located<string>;
  readonly proofPremises: Located<Premise>[];
  readonly proofConclusion: Located<PropositionH>;
};

export type Document = {
  readonly docTradition: Located<Tradition> | null;
  readonly docOpens: Located<string>[];
  readonly docProofs: Located<ProofBlock>[];
};

// ---------------------------------------------------------------------------
// fromConcreteH
// ---------------------------------------------------------------------------

export function fromConcreteH(ph: PropositionH): Proposition | null {
  if (
    ph.tag === "PropH" &&
    ph.propTypeH.tag === "ConcretePT" &&
    ph.subject.tag === "ConcreteT" &&
    ph.predicate.tag === "ConcreteT"
  ) {
    return {
      propType: ph.propTypeH.propType,
      subject: ph.subject.term,
      predicate: ph.predicate.term,
    };
  }
  return null;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function getPos(p: ParserState): SrcPos {
  return { posLine: p.line, posCol: p.col };
}

function located<T>(
  p: ParserState,
  parse: (p: ParserState) => T | null,
): Located<T> | null {
  const start = getPos(p);
  const value = parse(p);
  if (value === null) return null;
  const end = getPos(p);
  return { locStart: start, locEnd: end, locValue: value };
}

// Consume an end-of-line (or EOF). Returns true if consumed.
function eol(p: ParserState): boolean {
  if (p.eof) return true;
  return p.newline();
}

// ---------------------------------------------------------------------------
// Premise parsing
// ---------------------------------------------------------------------------

function refModifierP(p: ParserState): RefModifier | null {
  if (p.symbol("conv")) return "conv";
  if (p.symbol("per_accidens")) return "per_accidens";
  return null;
}

function refP(p: ParserState): Premise | null {
  if (p.peek() !== "@") return null;
  p.advance(); // consume '@'
  const first = p.takeWhile1(isNameChar, "proof name");
  if (first === null) return null;
  if (!p.eof && p.peek() === ".") {
    p.advance(); // consume '.'
    const name = p.takeWhile1(isNameChar, "proof name");
    if (name === null) return null;
    const modifier = refModifierP(p);
    return { tag: "PremiseRef", namespace: first, name, modifier };
  }
  const modifier = refModifierP(p);
  return { tag: "PremiseRef", namespace: null, name: first, modifier };
}

function premiseP(p: ParserState): Premise | null {
  // Try reference first
  const ref = refP(p);
  if (ref !== null) return ref;

  // Otherwise parse as proposition (possibly with holes)
  const ph = propositionHP(p);
  if (ph === null) return null;

  const concrete = fromConcreteH(ph);
  if (concrete !== null) return { tag: "PremiseProp", prop: concrete };
  return { tag: "PremiseHole", propH: ph };
}

// ---------------------------------------------------------------------------
// Tradition directive
// ---------------------------------------------------------------------------

function traditionP(p: ParserState): Tradition | null {
  if (!p.symbol("tradition")) return null;
  if (p.symbol("Strict")) return Tradition.Strict;
  if (p.symbol("Traditional")) return Tradition.Traditional;
  if (p.symbol("Full")) return Tradition.Full;
  return null;
}

// ---------------------------------------------------------------------------
// Open directive
// ---------------------------------------------------------------------------

function openP(p: ParserState): string | null {
  if (!p.symbol("open")) return null;
  return p.takeWhile1(isNameChar, "namespace name");
}

// ---------------------------------------------------------------------------
// Proof block
// ---------------------------------------------------------------------------

function conclusionP(p: ParserState): Located<PropositionH> | null {
  if (!p.symbol("∴") && !p.symbol("therefore")) return null;
  return located(p, propositionHP);
}

function proofBlockP(p: ParserState): ProofBlock | null {
  if (!p.symbol("proof")) return null;

  const name = located(p, (pp) => pp.takeWhile1(isNameChar, "proof name"));
  if (name === null) return null;

  if (!eol(p)) return null;
  p.scn();

  const premises: Located<Premise>[] = [];
  while (true) {
    const s0 = p.save();
    const prem = located(p, premiseP);
    if (prem === null) {
      p.restore(s0);
      break;
    }
    // Premise must be followed by EOL (or EOF before conclusion)
    if (!eol(p)) {
      p.restore(s0);
      break;
    }
    p.scn();
    premises.push(prem);
  }

  const concl = conclusionP(p);
  if (concl === null) return null;

  return { proofName: name, proofPremises: premises, proofConclusion: concl };
}

// ---------------------------------------------------------------------------
// Document parser
// ---------------------------------------------------------------------------

export function parseDocument(input: string): Document | ParseError {
  const p = new ParserState(input);
  p.scn();

  // Optional tradition directive
  let docTradition: Located<Tradition> | null = null;
  {
    const s0 = p.save();
    const trad = located(p, traditionP);
    if (trad !== null) {
      if (eol(p)) {
        p.scn();
        docTradition = trad;
      } else {
        p.restore(s0);
      }
    } else {
      p.restore(s0);
    }
  }

  // Open directives
  const docOpens: Located<string>[] = [];
  while (true) {
    const s0 = p.save();
    const op = located(p, openP);
    if (op !== null && eol(p)) {
      p.scn();
      docOpens.push(op);
    } else {
      p.restore(s0);
      break;
    }
  }

  // Proof blocks
  const docProofs: Located<ProofBlock>[] = [];
  while (!p.eof) {
    p.scn();
    if (p.eof) break;
    const proof = located(p, proofBlockP);
    if (proof === null) {
      return p.error("Expected 'proof' block");
    }
    p.scn();
    docProofs.push(proof);
  }

  return { docTradition, docOpens, docProofs };
}
